diff --git a/src/win32/cairo-win32-font.c b/src/win32/cairo-win32-font.c
index 9c5d8fc..747591c 100644
--- a/src/win32/cairo-win32-font.c
+++ b/src/win32/cairo-win32-font.c
@@ -322,7 +322,7 @@ get_first_monitor_scale_factor (HMONITOR hmonitor,
                                 LPARAM   data)
 {
     UINT dpi_x, dpi_y;
-    double *scale = (double *)data;
+    int *scale = (double *)data;
 
     if (!p_GetDpiForMonitor) {
         p_GetDpiForMonitor = (t_GetDpiForMonitor)GetProcAddress (GetModuleHandleA ("user32.dll"), "GetDpiForMonitorInternal");
@@ -336,12 +336,17 @@ get_first_monitor_scale_factor (HMONITOR hmonitor,
     return FALSE;
 }
 
-static double
+static int
 _win32_get_system_scale_factor (void)
 {
-    double scale = 1.0;
+    int scale = 1;
+	const char *scale_str = getenv ("GDK_SCALE");
 
-    EnumDisplayMonitors (NULL, NULL, get_first_monitor_scale_factor, (LPARAM)&scale);
+	if (scale_str) {
+		scale = atoi (scale_str);
+	} else {
+		EnumDisplayMonitors(NULL, NULL, get_first_monitor_scale_factor, (LPARAM)&scale);
+	}
 
     return scale;
 }
